File: src/decoders/instance.ts
import type { Instance } from "~/models";

export const decodeInstance = (instance: any): Instance => {
  let casURL: string | undefined;
  let casToken: string | undefined;

  if (instance.CAS.actif) {
    casURL = instance.CAS.casURL;
    casToken = instance.CAS.jetonCAS;
  }

  return {
    name: instance.nomEtab,
    version: instance.version,
    date: new Date(instance.date),
    accounts: instance.espaces.map((account: any) => ({
      name: account.nom,
      path: account.URL
    })),

    casURL, casToken
  };
};

File: library/src/commonMain/kotlin/decoders/Instance.kt
package decoders

import kotlinx.datetime.Instant
import kotlinx.serialization.json.*
import models.Instance
import models.InstanceAccount

fun decodeInstance(instance: JsonObject): Instance {
    var casURL: String? = null
    var casToken: String? = null

    if (instance["CAS"]!!.jsonObject["actif"]!!.jsonPrimitive.boolean) {
        casURL = instance["CAS"]!!.jsonObject["casURL"]!!.jsonPrimitive.content
        casToken = instance["CAS"]!!.jsonObject["jetonCAS"]!!.jsonPrimitive.content
    }

    return Instance(
        name = instance["nomEtab"]!!.jsonPrimitive.content,
        version = instance["version"]!!.jsonArray.map {version -> version.jsonPrimitive.int},
        date = Instant.parse(instance["date"]!!.jsonPrimitive.content),
        accounts = instance["espaces"]!!.jsonArray.map {account -> InstanceAccount(
            name = account.jsonObject["nom"]!!.jsonPrimitive.content,
            path = account.jsonObject["URL"]!!.jsonPrimitive.content
        ) },
        casURL = casURL,
        casToken = casToken
    )
}